hop/robust_publisher.py

```diff
@@ -0,0 +1,640 @@
+                       notifications. If this value is too large (larger than or similar in size to
+                       the time between messages needing to be sent) throughput will be lost as time
+                       will be spent waiting to see if the previous message has been acknowledged
+                       which could be better spent getting the next message sent out. When in doubt,
+                       it is probably best to err on the side of choosing a small value.
+            kwargs: Any additional arguments to be passed to :meth:`hop.io.open <hop.io.open>`.
+
+        Raises:
+            OSError: If a journal file exists but cannot be read.
+            Runtime Error: If the contents of the journal file are corrupted.
+        """
+        super(RobustPublisher, self).__init__()
+
+        self._poll_wait = poll_wait
+
+        # read any messages left over from a previous run
+        self._journal = PublicationJournal(journal_path)
+        if self._journal.has_messages_to_send():
+            logger.info(f"Journal has {len(self._journal.messages_to_send)} "
+                        "left-over messages to send")
+
+        self._lock = threading.Lock()
+        self._cond = threading.Condition(self._lock)
+
+        dummy = io.Stream(auth=auth)
+        self._stream = dummy.open(url, "w", error_callback=PublicationJournal.error_callback,
+                                  **kwargs)
+
+        self._running = True
+        self._immediate_start = self._journal.has_messages_to_send()
+
+    def run(self):
+        while True:
+            with self._cond:
+                if not self._running and not self._journal.has_messages_to_send():
+                    break
+                if not self._immediate_start:
+                    self._cond.wait()  # wait for something to do
+                else:
+                    self._immediate_start = False
+            self._do_send()
+        self._stream.flush()
+
+    def _do_send(self):
+        journal = self._journal
+        # use this slightly awkward loop structure to reacquire the lock on each iteration so that
+        # other threads have a chance to take it
+        while True:
+            with self._lock:
+                busy = journal.has_messages_to_send() or journal.has_messages_in_flight()
+                if not busy:
+                    break  # no work to do for now
+                do_send = False
+                if journal.has_messages_to_send():
+                    seq_num, message, headers = journal.get_next_message_to_send()
+                    do_send = True
+            if do_send:
+                try:
+                    dc = journal.get_delivery_callback(seq_num, self._lock)
+                    logger.debug(f"Sending message with sequence number {seq_num}")
+                    self._stream.write_raw(message, headers=headers, delivery_callback=dc)
+                except KafkaException as e:
```

Comment:
This appears to be redundant with the `get_delivery_callback` call a few lines above, which would also lead to a logger.error and message requeue (via `_delivery_callback`). Is this desired?

Reply:
I think that the unedrlying confluent-kafka-python function can fail in a number of ways before it could invoke a callback, for example if it has an argument parsing error, or for certain types of internal version mismatches in the C libraries.